6 dockerizing the application

bencefarkas200 opened the pull request.

bencefarkas200 committed:
Create Dockerfiles in front-end and back-end directories

bencefarkas200 committed:
Create Docker Image for the frontend

bencefarkas200 committed:
Include all package.json files in frontend Dockerfile

bencefarkas200 committed:
Create backend image

Also formalize frontend Dockerfile

bencefarkas200 committed:
Correct wrong cmd command

bencefarkas200 committed:
Create Docker Compose file

bencefarkas200 committed:
Implement docker-compose code, so Todo app can run as a muli-container app

Also made changes in backend Dockerfile (not starting in production mode anymore)

gy-szabo commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+      - backend
+
+  backend:
+    build: ./back-end
```
the backend build will fail in a new developer's machine, because the `.env` file is not included. This is not a problem, because we talked about not sharing that particular file. The `.env.samle` is a great way of communicating with new devs that you should fill these and these variables, but the main problem is that the sample file is not next to the .env file. This is misguiding a little, so please move the sample file where the .env file is located

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
```
Please manually format this file, so that you use only 0 or 1 blank line between commands, and divide the different steps with an extra blank line.

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+COPY . /app
+
+RUN npm run build --prod
```
We are going to use this app locally, therefore --prod build is unnecessary. Without the prod optimizations (AOT instead of JIT, tree shaking, uglifying, etc.), the ng build step will run faster

_Obviously we will need a production build to optimize our image for production use, but right now we are focusing on development only_

gy-szabo replied:
I see that you are using CMD now. Did `npm run build` not work?

gy-szabo commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+    ports:
+      - 3000:3000
+    environment:
+      DB_URL: mongodb://127.0.0.1:27017/todo/tasks
```
Well the values that we wanted to mask with the .env file are present here as well. I was thinking about setting these values from environment variables, but I'm not sure that we need this line even. Do we really need to set the `DB_URL` in the compose file?

gy-szabo replied:
I can see a probability that we do, and I also got a new idea. You should first merge the variables in the .env file into a DB_URL variable, we don't need 3 variables for the same cause anyways. Then (if you need) you could read the same `.env` file in the compose file as the environment. Just like here: https://docs.docker.com/compose/environment-variables/set-environment-variables/#use-the-env_file-attribute

gy-szabo commented on back-end/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
```
don't start with a blank line please

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+
+FROM node:latest as build-step
```
You are using `node:18` on the frontend side, but `node:latest` here.
1. Because both apps use node, you could use the same version for the build images
2. Latest is not often the best solution also. I'd suggest picking the highest available version at the time you are creating the images, but don't use the `latest`. Latest may be good for a long time, maybe forever, but it may lead to breaking changes in the future. _In short: pick concrete image version that is currently working with your app._

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+ 
+
+FROM nginx:1.25.3
```
I like that you gave a name for the build step. Given that, you should give this one a name as well to be consistent

gy-szabo commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+      - todo:/data/db
+
+volumes:
+  todo:
```
you are giving a name for your volumes here, and todo in itself is not enough. I'd suggest giving it a more intuitive name, that is distinguishable (either in docker desktop, or docker cli) from other volumes of other docker compose files. E.g. `todo_mongodb_data`

gy-szabo commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+    depends_on:
+      - database
+
+  database:
```
a good practice is to name your containers in a way that they are distinguishable from other containers. If you have another home project with some other name but with the same service name, that's quite confusing

_(e.g. you could have 2 containers called `backend`, but it would be better to have a `todo-backend` and a `other-home-project-backend` container)_

I'd suggest giving every service name a `todo-` prefix

bencefarkas200 replied:
Docker Desktop automaticly does that. And I think it's the same way in Docker CLI, too.

![image](https://github.com/bencefarkas200/to-do/assets/148462547/28ec91ba-e96c-4987-b950-10f23b0cb89b)

gy-szabo replied:
Indeed!

Can confirm that it works the same with the CLI too.

bencefarkas200 committed:
Move .env.sample file next to .env

Change host so it can be used as a container

bencefarkas200 committed:
Formalize Dockerfile

bencefarkas200 committed:
Change RUN command in Dockerfile

bencefarkas200 committed:
Implement requested changes in Dockerfiles

bencefarkas200 committed:
Change volume name so it's more intuitive

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -12,10 +10,9 @@ RUN npm install
 COPY . /app
 
-RUN npm run build --prod
+CMD [ "ng","build" ]
```
I see that you are using CMD now. Did `npm run build` withouth the `--prod` flag not work?

bencefarkas200 replied:
You're right, it also works with `npm run bulild`

bencefarkas200 committed:
Merge .env variables

Make necessary changes to backend so it can handle merged variables

bencefarkas200 committed:
Change CMD command to RUN command in frontend Dockerfile

gy-szabo commented on back-end/src/app.module.ts:
```diff
@@ -8,9 +8,7 @@ import { TaskService } from './task/task.service';
-    MongooseModule.forRoot(
-      `mongodb://${process.env.DB_HOST}:${process.env.DB_PORT}/${process.env.DB_NAME}`,
-    ),
+    MongooseModule.forRoot(<string>process.env.DB_URI),
```
In this line we (and also the mongoose `forRoot` function) want to avoid `undefined` or `null` values, which in a scenario where the someone did not set the environment variables correctly can definitely occur.

However, in the aforementioned scenario where the environment variable is undefined, with your code `<string>undefined` will still be `undefined` even after casting. This is bad, because with this method only the compile time "I only accept string" warning/error is resolved, but runtime it still can get an undefined value, that could lead to unexpected behaviour.

bencefarkas200 replied:
Would `MongooseModule.forRoot(<string>process.env.DB_URI ?? '')` be a solution for this problem?

gy-szabo replied:
`<string>undefined` will still be `undefined` , so the casting is unnecessary. Simply `(process.env.DB_URI ?? '')` is good

both reach the same goal, but if you use casting, you will have 1 more computation

bencefarkas200 committed:
Create a .env file for running the application in container

Now the application uses two different .env file. One for running in a container and one for running locally.
Also created a .env.sample file for the new .env file.

gy-szabo commented on docker-compose.yml:
```diff
@@ -12,8 +12,10 @@ services:
+      - .env
     environment:
-      DB_URL: mongodb://127.0.0.1:27017/todo/tasks
+      DB_URL: DB_URL
```
Since you have an env file, you don't need to specify `DB_URL` here

furthermore, you don't even have a `DB_URL` env variable set, _(you have `DB_URI`, which is the better naming btw as the mongoose documentation also calls it URI, kudos for that)_ so the `environment` section is completely unused and therefore unnecessary

bencefarkas200 committed:
Make changes to forRoot function, so 'undefined' or 'null' value is avoided

bencefarkas200 committed:
Remove unnecessary enviroment section from compose file

gy-szabo commented on back-end/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+
+COPY . .
+
+COPY .env ./
```
`COPY . .` already copies every file into the container. Is this command necessary?

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM node:18 as build-step
+
+RUN mkdir -p /app
```
`WORKDIR /app` will create the dir given it does not exist, therefore this is unnecessary

gy-szabo commented on front-end/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+RUN npm install
+
+COPY . /app
```
as `/app` is the workdir already, `COPY . .` should be here, as it does the same as `COPY . /app`

bencefarkas200 committed:
Implement requested changes to Dockerfiles
